// headroom-ai/dist/index.cjs
'use strict';

var chunkR6FB52AD_cjs = require('./chunk-R6FB52AD.cjs');

// src/simulate.ts
async function simulate(messages, options = {}) {
  const { client: providedClient, model, config, ...clientOptions } = options;
  const openaiMessages = chunkR6FB52AD_cjs.toOpenAI(messages);
  const client = providedClient ?? new chunkR6FB52AD_cjs.HeadroomClient(clientOptions);
  const body = {
    messages: openaiMessages,
    model: model ?? "gpt-4o",
    config: {
      default_mode: "simulate",
      generate_diff_artifact: true,
      ...config ? chunkR6FB52AD_cjs.deepSnakeCase(config) : {}
    }
  };
  const result = await client.compressRaw(body);
  return chunkR6FB52AD_cjs.deepCamelCase(result);
}

// src/shared-context.ts
var SharedContext = class {
  entries = /* @__PURE__ */ new Map();
  client;
  model;
  ttl;
  maxEntries;
  constructor(options = {}) {
    const { model, ttl, maxEntries, ...clientOptions } = options;
    this.client = new chunkR6FB52AD_cjs.HeadroomClient(clientOptions);
    this.model = model ?? "claude-sonnet-4-5-20250929";
    this.ttl = ttl ?? 3600;
    this.maxEntries = maxEntries ?? 100;
  }
  /**
   * Store content with compression.
   */
  async put(key, content, options) {
    this.evictExpired();
    this.evictIfFull();
    const messages = [{ role: "user", content }];
    let result;
    try {
      result = await this.client.compress(messages, { model: this.model });
    } catch {
      result = {
        messages,
        tokensBefore: content.length / 4,
        // rough estimate
        tokensAfter: content.length / 4,
        tokensSaved: 0,
        compressionRatio: 1,
        transformsApplied: [],
        ccrHashes: [],
        compressed: false
      };
    }
    const compressed = result.compressed ? typeof result.messages[0]?.content === "string" ? result.messages[0].content : JSON.stringify(result.messages[0]?.content ?? content) : content;
    const entry = {
      key,
      original: content,
      compressed,
      originalTokens: result.tokensBefore,
      compressedTokens: result.tokensAfter,
      agent: options?.agent ?? null,
      timestamp: Date.now() / 1e3,
      transforms: result.transformsApplied,
      savingsPercent: result.tokensBefore > 0 ? (result.tokensBefore - result.tokensAfter) / result.tokensBefore * 100 : 0
    };
    this.entries.set(key, entry);
    return entry;
  }
  /**
   * Get content by key. Returns compressed by default, full original if full=true.
   */
  get(key, options) {
    const entry = this.entries.get(key);
    if (!entry) return null;
    if (Date.now() / 1e3 - entry.timestamp > this.ttl) {
      this.entries.delete(key);
      return null;
    }
    return options?.full ? entry.original : entry.compressed;
  }
  /**
   * Get full entry metadata.
   */
  getEntry(key) {
    const entry = this.entries.get(key);
    if (!entry) return null;
    if (Date.now() / 1e3 - entry.timestamp > this.ttl) {
      this.entries.delete(key);
      return null;
    }
    return entry;
  }
  /**
   * List all stored keys (excluding expired).
   */
  keys() {
    this.evictExpired();
    return Array.from(this.entries.keys());
  }
  /**
   * Get aggregated statistics.
   */
  stats() {
    this.evictExpired();
    let totalOriginal = 0;
    let totalCompressed = 0;
    for (const entry of this.entries.values()) {
      totalOriginal += entry.originalTokens;
      totalCompressed += entry.compressedTokens;
    }
    const saved = totalOriginal - totalCompressed;
    return {
      entries: this.entries.size,
      totalOriginalTokens: totalOriginal,
      totalCompressedTokens: totalCompressed,
      totalTokensSaved: saved,
      savingsPercent: totalOriginal > 0 ? saved / totalOriginal * 100 : 0
    };
  }
  /**
   * Clear all entries.
   */
  clear() {
    this.entries.clear();
  }
  evictExpired() {
    const now = Date.now() / 1e3;
    for (const [key, entry] of this.entries) {
      if (now - entry.timestamp > this.ttl) {
        this.entries.delete(key);
      }
    }
  }
  evictIfFull() {
    while (this.entries.size >= this.maxEntries) {
      const oldest = this.entries.keys().next().value;
      if (oldest !== void 0) this.entries.delete(oldest);
    }
  }
};

// src/paths.ts
var HEADROOM_CONFIG_DIR_ENV = "HEADROOM_CONFIG_DIR";
var HEADROOM_WORKSPACE_DIR_ENV = "HEADROOM_WORKSPACE_DIR";
var HEADROOM_SAVINGS_PATH_ENV = "HEADROOM_SAVINGS_PATH";
var HEADROOM_TOIN_PATH_ENV = "HEADROOM_TOIN_PATH";
var HEADROOM_SUBSCRIPTION_STATE_PATH_ENV = "HEADROOM_SUBSCRIPTION_STATE_PATH";
function isNode() {
  return typeof process !== "undefined" && typeof process.versions !== "undefined" && typeof process.versions.node === "string";
}
function getEnv(name) {
  if (typeof process === "undefined" || !process.env) {
    return "";
  }
  const value = process.env[name];
  return typeof value === "string" ? value.trim() : "";
}
function homeDir() {
  if (!isNode()) {
    return "";
  }
  try {
    const os = chunkR6FB52AD_cjs.__require("os");
    if (os.homedir) {
      const home = os.homedir();
      if (home) return home;
    }
  } catch {
  }
  return getEnv("HOME") || getEnv("USERPROFILE") || "";
}
function joinPath(...parts) {
  if (!isNode()) {
    return parts.filter((p) => p !== "").join("/");
  }
  try {
    const path = chunkR6FB52AD_cjs.__require("path");
    if (path.join) return path.join(...parts.filter((p) => p !== ""));
  } catch {
  }
  const sep = process.platform === "win32" ? "\\" : "/";
  return parts.filter((p) => p !== "").join(sep);
}
function expandTilde(p) {
  if (!p.startsWith("~")) return p;
  const home = homeDir();
  if (!home) return p;
  if (p === "~") return home;
  if (p.startsWith("~/") || p.startsWith("~\\")) {
    return joinPath(home, p.slice(2));
  }
  return p;
}
function resolve(explicit, envVar, derived) {
  if (!isNode()) {
    return "";
  }
  if (explicit !== void 0 && explicit !== "") {
    return expandTilde(explicit);
  }
  const envValue = getEnv(envVar);
  if (envValue) {
    return expandTilde(envValue);
  }
  return derived;
}
function workspaceDir() {
  if (!isNode()) return "";
  const envValue = getEnv(HEADROOM_WORKSPACE_DIR_ENV);
  if (envValue) return expandTilde(envValue);
  const home = homeDir();
  if (!home) return "";
  return joinPath(home, ".headroom");
}
function configDir() {
  if (!isNode()) return "";
  const envValue = getEnv(HEADROOM_CONFIG_DIR_ENV);
  if (envValue) return expandTilde(envValue);
  const workspaceEnv = getEnv(HEADROOM_WORKSPACE_DIR_ENV);
  if (workspaceEnv) {
    return joinPath(expandTilde(workspaceEnv), "config");
  }
  const home = homeDir();
  if (!home) return "";
  return joinPath(home, ".headroom", "config");
}
function savingsPath(explicit) {
  return resolve(
    explicit,
    HEADROOM_SAVINGS_PATH_ENV,
    joinPath(workspaceDir(), "proxy_savings.json")
  );
}
function toinPath(explicit) {
  return resolve(
    explicit,
    HEADROOM_TOIN_PATH_ENV,
    joinPath(workspaceDir(), "toin.json")
  );
}
function subscriptionStatePath(explicit) {
  return resolve(
    explicit,
    HEADROOM_SUBSCRIPTION_STATE_PATH_ENV,
    joinPath(workspaceDir(), "subscription_state.json")
  );
}
function memoryDbPath() {
  if (!isNode()) return "";
  return joinPath(workspaceDir(), "memory.db");
}
function nativeMemoryDir() {
  if (!isNode()) return "";
  return joinPath(workspaceDir(), "memories");
}
function licenseCachePath() {
  if (!isNode()) return "";
  return joinPath(workspaceDir(), "license_cache.json");
}
function sessionStatsPath() {
  if (!isNode()) return "";
  return joinPath(workspaceDir(), "session_stats.jsonl");
}
function syncStatePath() {
  if (!isNode()) return "";
  return joinPath(workspaceDir(), "sync_state.json");
}
function bridgeStatePath() {
  if (!isNode()) return "";
  return joinPath(workspaceDir(), "bridge_state.json");
}
function logDir() {
  if (!isNode()) return "";
  return joinPath(workspaceDir(), "logs");
}
function proxyLogPath() {
  if (!isNode()) return "";
  return joinPath(logDir(), "proxy.log");
}
function debug400Dir() {
  if (!isNode()) return "";
  return joinPath(logDir(), "debug_400");
}
function binDir() {
  if (!isNode()) return "";
  return joinPath(workspaceDir(), "bin");
}
function rtkPath() {
  if (!isNode()) return "";
  const name = process.platform === "win32" ? "rtk.exe" : "rtk";
  return joinPath(binDir(), name);
}
function deployRoot() {
  if (!isNode()) return "";
  return joinPath(workspaceDir(), "deploy");
}
function beaconLockPath(port) {
  if (!isNode()) return "";
  return joinPath(workspaceDir(), `.beacon_lock_${Math.trunc(port)}`);
}
function modelsConfigPath() {
  if (!isNode()) return "";
  return joinPath(configDir(), "models.json");
}
function assertPluginName(name) {
  if (!name || name.includes("/") || name.includes("\\")) {
    throw new Error(`invalid plugin name: ${JSON.stringify(name)}`);
  }
}
function pluginConfigDir(pluginName) {
  assertPluginName(pluginName);
  if (!isNode()) return "";
  return joinPath(configDir(), "plugins", pluginName);
}
function pluginWorkspaceDir(pluginName) {
  assertPluginName(pluginName);
  if (!isNode()) return "";
  return joinPath(workspaceDir(), "plugins", pluginName);
}

Object.defineProperty(exports, "CacheError", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.CacheError; }
});
Object.defineProperty(exports, "CompressionHooks", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.CompressionHooks; }
});
Object.defineProperty(exports, "ConfigurationError", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.ConfigurationError; }
});
Object.defineProperty(exports, "HeadroomAuthError", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.HeadroomAuthError; }
});
Object.defineProperty(exports, "HeadroomClient", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.HeadroomClient; }
});
Object.defineProperty(exports, "HeadroomCompressError", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.HeadroomCompressError; }
});
Object.defineProperty(exports, "HeadroomConnectionError", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.HeadroomConnectionError; }
});
Object.defineProperty(exports, "HeadroomError", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.HeadroomError; }
});
Object.defineProperty(exports, "ProviderError", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.ProviderError; }
});
Object.defineProperty(exports, "StorageError", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.StorageError; }
});
Object.defineProperty(exports, "TokenizationError", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.TokenizationError; }
});
Object.defineProperty(exports, "TransformError", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.TransformError; }
});
Object.defineProperty(exports, "ValidationError", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.ValidationError; }
});
Object.defineProperty(exports, "camelToSnake", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.camelToSnake; }
});
Object.defineProperty(exports, "collectStream", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.collectStream; }
});
Object.defineProperty(exports, "compress", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.compress; }
});
Object.defineProperty(exports, "countTurns", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.countTurns; }
});
Object.defineProperty(exports, "deepCamelCase", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.deepCamelCase; }
});
Object.defineProperty(exports, "deepSnakeCase", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.deepSnakeCase; }
});
Object.defineProperty(exports, "detectFormat", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.detectFormat; }
});
Object.defineProperty(exports, "extractToolCalls", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.extractToolCalls; }
});
Object.defineProperty(exports, "extractUserQuery", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.extractUserQuery; }
});
Object.defineProperty(exports, "fromOpenAI", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.fromOpenAI; }
});
Object.defineProperty(exports, "mapProxyError", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.mapProxyError; }
});
Object.defineProperty(exports, "parseSSE", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.parseSSE; }
});
Object.defineProperty(exports, "snakeToCamel", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.snakeToCamel; }
});
Object.defineProperty(exports, "toOpenAI", {
  enumerable: true,
  get: function () { return chunkR6FB52AD_cjs.toOpenAI; }
});
exports.HEADROOM_CONFIG_DIR_ENV = HEADROOM_CONFIG_DIR_ENV;
exports.HEADROOM_SAVINGS_PATH_ENV = HEADROOM_SAVINGS_PATH_ENV;
exports.HEADROOM_SUBSCRIPTION_STATE_PATH_ENV = HEADROOM_SUBSCRIPTION_STATE_PATH_ENV;
exports.HEADROOM_TOIN_PATH_ENV = HEADROOM_TOIN_PATH_ENV;
exports.HEADROOM_WORKSPACE_DIR_ENV = HEADROOM_WORKSPACE_DIR_ENV;
exports.SharedContext = SharedContext;
exports.beaconLockPath = beaconLockPath;
exports.binDir = binDir;
exports.bridgeStatePath = bridgeStatePath;
exports.configDir = configDir;
exports.debug400Dir = debug400Dir;
exports.deployRoot = deployRoot;
exports.licenseCachePath = licenseCachePath;
exports.logDir = logDir;
exports.memoryDbPath = memoryDbPath;
exports.modelsConfigPath = modelsConfigPath;
exports.nativeMemoryDir = nativeMemoryDir;
exports.pluginConfigDir = pluginConfigDir;
exports.pluginWorkspaceDir = pluginWorkspaceDir;
exports.proxyLogPath = proxyLogPath;
exports.rtkPath = rtkPath;
exports.savingsPath = savingsPath;
exports.sessionStatsPath = sessionStatsPath;
exports.simulate = simulate;
exports.subscriptionStatePath = subscriptionStatePath;
exports.syncStatePath = syncStatePath;
exports.toinPath = toinPath;
exports.workspaceDir = workspaceDir;
