// headroom-ai/dist/adapters/anthropic.cjs
'use strict';

var chunkR6FB52AD_cjs = require('../chunk-R6FB52AD.cjs');

// src/adapters/anthropic.ts
function anthropicToOpenAI(messages) {
  const result = [];
  for (const msg of messages) {
    if (msg.role === "user") {
      if (typeof msg.content === "string") {
        result.push({ role: "user", content: msg.content });
      } else if (Array.isArray(msg.content)) {
        const toolResults = msg.content.filter(
          (b) => b.type === "tool_result"
        );
        const textBlocks = msg.content.filter((b) => b.type === "text");
        if (textBlocks.length > 0) {
          result.push({
            role: "user",
            content: textBlocks.map((b) => b.text).join("\n")
          });
        }
        for (const tr of toolResults) {
          result.push({
            role: "tool",
            content: typeof tr.content === "string" ? tr.content : JSON.stringify(tr.content),
            tool_call_id: tr.tool_use_id
          });
        }
      }
      continue;
    }
    if (msg.role === "assistant") {
      if (typeof msg.content === "string") {
        result.push({ role: "assistant", content: msg.content });
      } else if (Array.isArray(msg.content)) {
        const textBlocks = msg.content.filter((b) => b.type === "text");
        const toolUseBlocks = msg.content.filter(
          (b) => b.type === "tool_use"
        );
        const content = textBlocks.length > 0 ? textBlocks.map((b) => b.text).join("\n") : null;
        const openaiMsg = { role: "assistant", content };
        if (toolUseBlocks.length > 0) {
          openaiMsg.tool_calls = toolUseBlocks.map(
            (b) => ({
              id: b.id,
              type: "function",
              function: {
                name: b.name,
                arguments: JSON.stringify(b.input)
              }
            })
          );
        }
        result.push(openaiMsg);
      }
      continue;
    }
  }
  return result;
}
function openAIToAnthropic(messages) {
  const result = [];
  for (const msg of messages) {
    if (msg.role === "user") {
      if (typeof msg.content === "string") {
        result.push({ role: "user", content: msg.content });
      } else if (Array.isArray(msg.content)) {
        result.push({
          role: "user",
          content: msg.content.map((p) => {
            if (p.type === "text") return { type: "text", text: p.text };
            return { type: "text", text: "" };
          })
        });
      }
      continue;
    }
    if (msg.role === "assistant") {
      const blocks = [];
      if (msg.content) blocks.push({ type: "text", text: msg.content });
      if (msg.tool_calls) {
        for (const tc of msg.tool_calls) {
          blocks.push({
            type: "tool_use",
            id: tc.id,
            name: tc.function.name,
            input: JSON.parse(tc.function.arguments)
          });
        }
      }
      result.push({
        role: "assistant",
        content: blocks.length === 1 && blocks[0].type === "text" ? blocks[0].text : blocks
      });
      continue;
    }
    if (msg.role === "tool") {
      result.push({
        role: "user",
        content: [
          {
            type: "tool_result",
            tool_use_id: msg.tool_call_id,
            content: msg.content
          }
        ]
      });
      continue;
    }
    if (msg.role === "system") {
      result.push({ role: "user", content: msg.content });
    }
  }
  return result;
}
function withHeadroom(client, options = {}) {
  const originalCreate = client.messages.create.bind(client.messages);
  const compressedCreate = async (params) => {
    const messages = params.messages;
    const model = options.model ?? params.model ?? "claude-sonnet-4-5-20250929";
    const openaiMessages = anthropicToOpenAI(messages);
    const result = await chunkR6FB52AD_cjs.compress(openaiMessages, {
      stack: "adapter_ts_anthropic",
      ...options,
      model
    });
    const anthropicMessages = result.compressed ? openAIToAnthropic(result.messages) : messages;
    return originalCreate({
      ...params,
      messages: anthropicMessages
    });
  };
  const messagesProxy = new Proxy(client.messages, {
    get(target, prop) {
      if (prop === "create") return compressedCreate;
      return target[prop];
    }
  });
  return new Proxy(client, {
    get(target, prop) {
      if (prop === "messages") return messagesProxy;
      return target[prop];
    }
  });
}

exports.withHeadroom = withHeadroom;
