// headroom-ai/dist/adapters/vercel-ai.js
import { vercelToOpenAI, compress, openAIToVercel } from '../chunk-7AJIWIPP.js';
import { createRequire } from 'module';

function headroomMiddleware(options = {}) {
  return {
    transformParams: async ({
      params
    }) => {
      const prompt = params.prompt;
      if (!prompt || prompt.length === 0) return params;
      const model = options.model ?? params.modelId ?? "gpt-4o";
      const openaiMessages = vercelToOpenAI(prompt);
      const result = await compress(openaiMessages, {
        stack: "adapter_ts_vercel_ai",
        ...options,
        model
      });
      if (!result.compressed) return params;
      const compressedPrompt = openAIToVercel(result.messages);
      return { ...params, prompt: compressedPrompt };
    }
  };
}
async function compressVercelMessages(messages, options = {}) {
  const openaiMessages = vercelToOpenAI(messages);
  const result = await compress(openaiMessages, {
    stack: "adapter_ts_vercel_ai",
    ...options
  });
  const vercelMessages = openAIToVercel(result.messages);
  return {
    ...result,
    messages: vercelMessages
  };
}
function withHeadroom(model, options = {}) {
  let wrapLanguageModel;
  try {
    const require2 = createRequire(import.meta.url);
    const ai = require2("ai");
    wrapLanguageModel = ai.wrapLanguageModel;
  } catch {
    throw new Error(
      'withHeadroom() requires the "ai" package. Install it with: npm install ai'
    );
  }
  return wrapLanguageModel({
    model,
    middleware: headroomMiddleware(options)
  });
}

export { compressVercelMessages, headroomMiddleware, withHeadroom };
