// headroom-ai/dist/adapters/vercel-ai.cjs
'use strict';

var chunkR6FB52AD_cjs = require('../chunk-R6FB52AD.cjs');
var module$1 = require('module');

var _documentCurrentScript = typeof document !== 'undefined' ? document.currentScript : null;
function headroomMiddleware(options = {}) {
  return {
    transformParams: async ({
      params
    }) => {
      const prompt = params.prompt;
      if (!prompt || prompt.length === 0) return params;
      const model = options.model ?? params.modelId ?? "gpt-4o";
      const openaiMessages = chunkR6FB52AD_cjs.vercelToOpenAI(prompt);
      const result = await chunkR6FB52AD_cjs.compress(openaiMessages, {
        stack: "adapter_ts_vercel_ai",
        ...options,
        model
      });
      if (!result.compressed) return params;
      const compressedPrompt = chunkR6FB52AD_cjs.openAIToVercel(result.messages);
      return { ...params, prompt: compressedPrompt };
    }
  };
}
async function compressVercelMessages(messages, options = {}) {
  const openaiMessages = chunkR6FB52AD_cjs.vercelToOpenAI(messages);
  const result = await chunkR6FB52AD_cjs.compress(openaiMessages, {
    stack: "adapter_ts_vercel_ai",
    ...options
  });
  const vercelMessages = chunkR6FB52AD_cjs.openAIToVercel(result.messages);
  return {
    ...result,
    messages: vercelMessages
  };
}
function withHeadroom(model, options = {}) {
  let wrapLanguageModel;
  try {
    const require2 = module$1.createRequire((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.tagName.toUpperCase() === 'SCRIPT' && _documentCurrentScript.src || new URL('vercel-ai.cjs', document.baseURI).href)));
    const ai = require2("ai");
    wrapLanguageModel = ai.wrapLanguageModel;
  } catch {
    throw new Error(
      'withHeadroom() requires the "ai" package. Install it with: npm install ai'
    );
  }
  return wrapLanguageModel({
    model,
    middleware: headroomMiddleware(options)
  });
}

exports.compressVercelMessages = compressVercelMessages;
exports.headroomMiddleware = headroomMiddleware;
exports.withHeadroom = withHeadroom;
